src/main/java/com/example/dog/service/TransactionalDogService.java

```diff
@@ -0,0 +1,57 @@
+package com.example.dog.service;
+
+import com.example.dog.dao.JdbcConnectionHolder;
+import com.example.dog.dao.ObjectNotFoundException;
+import com.example.dog.model.Dog;
+
+public class TransactionalDogService implements DogService {
+
+    private final JdbcDogService dogService;
```

Comment:
Also it's better to use the interface. This way you can implement yet another proxy (e.g. for logging or some other weird reason) and wrap the main implementation with a proxy, and then wrap that proxy with another proxy.

In general whatever code you write - the less the code knows about the injected object the better. So if working with interface is enough - work with it. Same way as when we declare a variable `List` instead of `ArrayList`